content/learning-paths/smartphones-and-mobile/ray_tracing/rt06_reflections.md

```diff
@@ -0,0 +1,439 @@
+    vec4 hit_base_color = get_bindless_base_color(material_id, hit_uv, texture_mip);
+
+    illuminated_hit_color = get_pbr_color(hit_pos, hit_normal, ray_orig, hit_base_color, hit_emissive, hit_material_properties);
+}
+
+void process_reflection_rq_hit(rayQueryEXT rayQuery, out vec4 illuminated_hit_color)
+{
+    vec2 hit_uv;
+    uint material_id;
+    vec4 hit_material_properties;
+    vec3 hit_pos;
+    vec3 hit_normal;
+    obtain_rq_hit_data(rayQuery, hit_material_properties, hit_pos, hit_normal, hit_uv, material_id);
+    illuminate_hit_data(hit_material_properties, hit_pos, hit_normal, hit_uv, material_id, illuminated_hit_color);
+}
+
+void launch_reflection_ray(vec3 ray_orig, vec3 incident_dir, vec3 normal, out vec4 reflection_color)
+{
+    const vec3 ray_dir = reflect(incident_dir, normal);
+    rayQueryEXT rayQuery;
+    bool valid_hit = trace_ray(ray_orig, ray_dir, rayQuery);
+    if (valid_hit)
+    {
+        // Illuminate hit using PBR
+        process_reflection_rq_hit(rayQuery, reflection_color);
+    }
+    else
+    {
+        // Ray miss: Sample the sky
+        reflection_color.rgb = textureLod(environment_map, ray_dir, 0).rgb;
+        reflection_color.a = 1.0;
+    }
+}
+
+vec4 main()
+{
+    o_color = vec4(0);
+    // Retrieve some data from the G-buffer
+    vec2 g_buffer_uv = vec2(gl_FragCoord.xy) / current_resolution;
+
+    ivec2 texture_size = textureSize(depth_sampler, 0);
+    ivec2 depth_coord_texel = ivec2(g_buffer_uv * texture_size);
+    float depth = get_g_buffer_depth(depth_coord_texel);
+    if (depth <= 0)
+        return;
+
+    const vec3 pos = get_g_buffer_world_pos(g_buffer_uv);
+    const vec4 material_properties = get_g_buffer_material_properties(g_buffer_uv);
+    const vec3 normal = get_g_buffer_normal(g_buffer_uv);
+    const bool is_reflective = get_material_is_reflective(material_properties);
+
+    bool is_valid_ray = is_reflective;
+    // Reflection ray
+    if (is_valid_ray)
+    {
+        const vec3 incident_dir = normalize(pos - camera_position);
+        vec4 reflection_color = vec4(0);
+        launch_reflection_ray(pos, incident_dir, normal, reflection_color);
+
+        o_color += reflection_color;
+    }
+}
+```
+
+### Glossy reflections
+
+When creating our rays, we can use the roughness to decide how many reflection rays we need to launch.
+
+When we think of reflections, we usually think of mirror like reflections, however non mirror objects also reflect part of their incident light producing glossy or rough reflections. In a PBR render we use the material roughness to decide the reflected light dispersion. A roughness of 0 will indicate that the material is a mirror, so we will need a single ray to solve it. On the other hand, objects with a higher roughness will reflect the incident ray in multiple directions, producing glossy reflections. A render will need to launch multiple rays to solve the color of glossy reflections.
```

Comment:
hyphens (and there probably could have been a few I've not called out before now...): it should be "mirror-like" and "non-mirror"

Reply:
I tried to add all the ones I found.